### ISSUE TITLE
Unable to compile gdal with Visual Studio 2017

### ISSUE THREAD

**Issue by OgreTransporter** [OSGeo/gdal#2308]
## Expected behavior and actual behavior.

I'm not able to compile a debug Version of GDAL with MS Visual Studio 2017 Community. Here only 2 of the 358 same error message with different objects in PROJ:

```
proj_d.lib(4D_api.obj) : error LNK2038: Conflict determined for "_ITERATOR_DEBUG_LEVEL": The value "2" does not match the value "0" in cplgetsymbol.obj.
proj_d.lib(4D_api.obj) : error LNK2038: Conflict detected for "RuntimeLibrary": The value "MDd_DynamicDebug" does not match the value "MD_DynamicRelease" in cplgetsymbol.obj.
```

Building a debug version of GDAL does not seem to work. At least the error message says that cplgetsymbol.obj is compiled as release, which of course leads to an error when linking with the debug version of PROJ.

## Steps to reproduce the problem.
```
SET WORKSPACE=D:\dev\workspace
cd /d %WORKSPACE%
git clone https://github.com/OSGeo/gdal.git src
call "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Auxiliary\Build\vcvarsall.bat"  x64
mkdir %WORKSPACE%\build
mkdir %WORKSPACE%\install
mkdir %WORKSPACE%\install-temp
mkdir %WORKSPACE%\install-temp\Debug
mkdir %WORKSPACE%\install-temp\Debug\bin
mkdir %WORKSPACE%\install-temp\Release
mkdir %WORKSPACE%\install-temp\Release\bin
```
Then the configuration file (D:\dev\workspace\src\gdal\nmake.local):
```
!IFNDEF MSVC_VER
MSVC_VER=1916
!ENDIF

VERSION = 310
SWIG = D:\dev\Tools\swigwin-4.0.1\swig.exe
PYDIR = D:\dev\Tools\WinPython-3.6.8.0\python-3.6.8.amd64 
SPECIAL_DEPENDENCIES = D:\dev\workspace\deps
!IF "$(DEBUG)" == "1"
GDAL_HOME = D:\dev\workspace\install-temp\Debug
!ELSE
GDAL_HOME = D:\dev\workspace\install-temp\Release
!ENDIF

EXPAT_DIR = $(SPECIAL_DEPENDENCIES)\expat\install
EXPAT_INCLUDE = -I$(EXPAT_DIR)/include /DXML_STATIC
!IF "$(DEBUG)" == "1"
EXPAT_LIB = $(EXPAT_DIR)\lib\libexpat_staticd.lib
!ELSE
EXPAT_LIB = $(EXPAT_DIR)\lib\libexpat_static.lib
!ENDIF

XERCES_DIR = $(SPECIAL_DEPENDENCIES)\Xerces-C++\install
XERCES_INCLUDE = -I$(XERCES_DIR)/include -I$(XERCES_DIR)/include/xercesc
!IF "$(DEBUG)" == "1"
XERCES_LIB = $(XERCES_DIR)\lib\xerces-c_3D.lib Advapi32.lib
!ELSE
XERCES_LIB = $(XERCES_DIR)\lib\xerces-c_3.lib Advapi32.lib
!ENDIF

ILI_ENABLED = YES

SQLITE_DIR = $(SPECIAL_DEPENDENCIES)\sqlite3\install
SQLITE_INC = -I$(SQLITE_DIR)\src
!IFNDEF DEBUG
SQLITE_LIB = $(SQLITE_DIR)\lib\Release\sqlite_static.lib
!ELSE
SQLITE_LIB = $(SQLITE_DIR)\lib\Debug\sqlite_staticd.lib
!ENDIF

PROJ_DIR = $(SPECIAL_DEPENDENCIES)\PROJ\install
PROJ_FLAGS = -DPROJ_STATIC
PROJ_INCLUDE = -I$(PROJ_DIR)\include
!IF "$(DEBUG)" == "1"
PROJ_LIBRARY = $(PROJ_DIR)\lib_static\proj_d.lib Shell32.lib
!ELSE
PROJ_LIBRARY = $(PROJ_DIR)\lib_static\proj.lib Shell32.lib
!ENDIF

CURL_DIR = $(SPECIAL_DEPENDENCIES)\curl\install
CURL_INC = -I$(CURL_DIR)/include
!IF "$(DEBUG)" == "1"
CURL_LIB = $(CURL_DIR)/lib/libcurld_imp.lib wsock32.lib wldap32.lib winmm.lib
!ELSE
CURL_LIB = $(CURL_DIR)/lib/libcurl_imp.lib wsock32.lib wldap32.lib winmm.lib
!ENDIF

ZLIB_DIR = $(SPECIAL_DEPENDENCIES)\zlib\install
ZLIB_EXTERNAL_LIB=1
ZLIB_INC = -I$(ZLIB_DIR)\include
!IF "$(DEBUG)" == "1"
ZLIB_LIB = $(ZLIB_DIR)\lib\zlibstaticd.lib
!ELSE
ZLIB_LIB = $(ZLIB_DIR)\lib\zlibstatic.lib
!ENDIF

PCRE_DIR = $(SPECIAL_DEPENDENCIES)\PCRE\install
PCRE_INC=-I$(PCRE_DIR)\include -DHAVE_PCRE -DPCRE_STATIC
!IF "$(DEBUG)" == "1"
PCRE_LIB=$(PCRE_DIR)\lib\pcred.lib
!ELSE
PCRE_LIB=$(PCRE_DIR)\lib\pcre.lib
!ENDIF

LIBXML2_DIR = $(SPECIAL_DEPENDENCIES)\libxml2\install
LIBXML2_INC = -I$(LIBXML2_DIR)\include\libxml2
!IF "$(DEBUG)" == "1"
LIBXML2_LIB = $(LIBXML2_DIR)\lib\Debug\libxml2.lib
!ELSE
LIBXML2_LIB = $(LIBXML2_DIR)\lib\Release\libxml2.lib
!ENDIF

HDF5_PLUGIN = YES
HDF5_DIR = $(SPECIAL_DEPENDENCIES)\hdf5\install
!IF "$(DEBUG)" == "1"
HDF5_LIB = $(HDF5_DIR)\lib\libhdf5_D.lib
!ELSE
HDF5_LIB = $(HDF5_DIR)\lib\libhdf5.lib
!ENDIF

WEBP_DIR = $(SPECIAL_DEPENDENCIES)\libwebp\install
WEBP_ENABLED = YES
WEBP_CFLAGS = -I$(WEBP_DIR)\Include
!IF "$(DEBUG)" == "1"
WEBP_LIBS = $(WEBP_DIR)\lib\webpd.lib
!ELSE
WEBP_LIBS = $(WEBP_DIR)\lib\webp.lib
!ENDIF

BINDINGS=csharp python
HAVE_NUMPY=1
```
```
nmake -f makefile.vc MSVC_VER=1916 DEBUG=1 WITH_PDB=1 GDAL_HOME=D:\Jenkins\workspace\install-temp\Debug WIN64=1
```

## Operating system

Windows 10 + Visual Studio 2017 Community

## GDAL version and provenance

```
commit 0d56970161f91aa66441c0b2166ca5b23ce3382d (HEAD -> master, origin/master, origin/HEAD)
Merge: 4ec6d1c89d a68b4b9092
Author: Even Rouault <even.rouault@spatialys.com>
Date:   Mon Mar 9 20:58:17 2020 +0100

    Merge pull request #2296 from mapycz/swift-auth-v3

    /vsiswift/: V3 authentication method, handling auth token expiration
```

## Comment

I know the issue has been raised and discarded many times, but a CMake build would simplify things. At least on Windows it is easier to debug CMake scripts than makefiles.

**Comment by rouault** [comment]
It looks like the CRT used by GDAL and PROJ is not the same. GDAL uses /MD whereas PROJ likely uses /MDd. 
But GDAL ought to use /MDd . This was fixes in 3.0 or something, but I just found this was accidentaly reverted in https://github.com/OSGeo/gdal/commit/5c1f31776c74aa8256c253f0c4915109a605bf3f#diff-742e1ef88f8f0fe54e11f25d0b6e75d0R139 . So I'm going to re-revert that part of the commit to use /MDd.

**Comment by OgreTransporter** [comment]
/MDd is correct for debug


Option | Description
-- | --
/MD | Causes the application to use the multithread-specific and DLL-specific version of the run-time library. Defines _MT and _DLL and causes the compiler to place the library name MSVCRT.lib into the .obj file. Applications compiled with this option are statically linked to MSVCRT.lib. This library provides a layer of code that enables the linker to resolve external references. The actual working code is contained in MSVCRversionnumber.DLL, which must be available at run time to applications linked with MSVCRT.lib.
/MDd | Defines _DEBUG, _MT, and _DLL and causes the application to use the debug multithread-specific and DLL-specific version of the run-time library. It also causes the compiler to place the library name MSVCRTD.lib into the .obj file.

Source: https://docs.microsoft.com/en-us/cpp/build/reference/md-mt-ld-use-run-time-library?view=vs-2017
